Add loading tags and code (pandas, dask, wds)

lhoestq opened the pull request:
I added example codes to load with libraries:
- mlcroissant
- datasets
- pandas or dask or webdataset

The new associated `libraries` tags are: "pandas", "dask", "datasets", "mlcroissant", "webdataset"

I also added a new field `loading_codes`. It is a list (currently with maximum three methods) of loading methods in python.

A loading method is
- a library name (datasets/pandas/dask/webdataset/mlcroissant)
- a function name
- a list of loading codes (one per config)

A loading code is
- a config_name
- a dict of arguments
- a code example

related to https://github.com/huggingface/datasets-server/issues/2414

julien-c commented on services/worker/src/worker/job_runners/dataset/loading_tags.py:
```diff
@@ -2,24 +2,558 @@
+    for pattern in datasets.data_files.KEYWORDS_IN_PATH_NAME_BASE_PATTERNS
+):
+    raise ImportError(
+        f"Current `datasets` version is not comptible with simplify_data_files_patterns which expects as separator {NON_WORD_GLOB_SEPARATOR}. "
```
```suggestion
        f"Current `datasets` version is not compatible with simplify_data_files_patterns which expects as separator {NON_WORD_GLOB_SEPARATOR}. "
```

severo reviewed (commented):
Super impressive.

I think it must be complemented with a series of unit tests to ensure all the cases work as expected.
Possibly in the tests we can even run the generated code to be sure it does what is needed.

lhoestq commented on services/worker/src/worker/job_runners/dataset/loading_tags.py:
```diff
@@ -41,19 +575,35 @@ def compute_loading_tags_response(dataset: str) -> DatasetLoadingTagsResponse:
+                if python_loading_method["library"] in loading_tag_for_library:
+                    tags.append(loading_tag_for_library[python_loading_method["library"]])
+                python_loading_methods.append(python_loading_method)
+            except (NotImplementedError, FileNotFoundError):
```
I should remove it, good catch

lhoestq commented on services/worker/src/worker/job_runners/dataset/loading_tags.py:
```diff
@@ -41,19 +575,35 @@ def compute_loading_tags_response(dataset: str) -> DatasetLoadingTagsResponse:
+        builder_name = infos[0]["builder_name"]
+        if builder_name in get_python_loading_method_for_builder:
+            try:
+                python_loading_method = get_python_loading_method_for_builder[builder_name](dataset, hf_token)
```
Currently we return only one builder-specific loading method: pandas OR dask OR webdataset

This loading method is added to the list of loading methods that can also include `datasets` and `mlcroissant`

severo replied:
OK. So: why not return an array so that if we change our mind or want to return two values for a specific case, we won't have to change the code in moonlanding?

lhoestq replied:
The jobs does return an array ;)

```python
python_loading_methods: list[PythonLoadingMethod]
```

severo replied:
wow ok, sorry about that :)

lhoestq commented:
Took your comments into account and added a bunch of tests @severo :)

severo reviewed (approved):
yeah!

severo commented:
hmmm: before merging, @julien-c what do you think of the tags? should we prefix with `python-library:` to be more precise and to make it easier to classify on moonlanding?

severo commented on services/worker/src/worker/job_runners/dataset/loading_tags.py:
```diff
@@ -2,24 +2,558 @@
+            {
+                "config_name": info["config_name"],
+                "arguments": {
+                    "record_set": info["config_name"]
```
@lhoestq you could reuse this function in https://github.com/huggingface/datasets-server/blob/aca4fc097a594ad0d58e338a30c478a05ce6acf7/services/api/src/api/routes/croissant.py#L141

lhoestq commented:
I don't see the point of prefixing tags, it makes it ugly/hard to read no ?

severo commented:
> I don't see the point of prefixing tags, it makes it ugly/hard to read no ?

the UI is not a matter, we show what we want (we can strip the prefix). It's more a matter of giving the Hub a way to classify these tags as python libraries.

severo commented on libs/libcommon/src/libcommon/croissant_utils.py:
nit: copyright header

lhoestq commented:
Did some more changes:
- moved the libraries to "libraries" instead of the flat list of "tags".
- kept the tag `croissant` in "tags"
- added `mlcroissant` to "libraries"
- moved `webdataset` from "tags" to "libraries"

lmk what you think @severo

severo commented on services/worker/src/worker/dtos.py:
```diff
@@ -218,19 +218,37 @@ class IsValidResponse(TypedDict):
 class DatasetLoadingTagsResponse(TypedDict):
-    tags: list[DatasetLoadingTag]
+    tags: list[DatasetTag]
+    libraries: list[DatasetLibrary]
```
isn't libraries redundant with loading_methods, since we get the same info? maybe merge them

severo reviewed (approved):
OK to me. I think we can simply remove "libraries" since we have the same info in "loading_methods"; and if so, we can rename "loading_methods" into "libraries" for clarity
